core/shared/src/main/scala/zio/package.scala

```diff
@@ -42,6 +42,16 @@ package object zio extends EitherCompat with PlatformSpecific with VersionSpecif
   type Ref[A]      = ZRef[Nothing, Nothing, A, A]
   type ERef[+E, A] = ZRef[E, E, A, A]
 
+  type ZLens[-S, +T, +A, -B]      = (S => A, B => S => T)
+  type ZOptional[-S, +T, +A, -B]  = (S => Option[A], B => S => T)
+  type ZPrism[-S, +T, +A, -B]     = (S => Option[A], B => T)
+  type ZTraversal[-S, +T, +A, -B] = (S => List[A], List[B] => S => Option[T])
```

Comment:
I'm encoding a traversal right now as a pair of functions `S => List[A]` and `List[B] => S => Option[T]`. The set operation can fail if the number of elements being set doesn't match the expected number of elements to be set in the structure.